File: redis_files/redis_calculations.py
"""High-performance Redis-backed technical indicator calculations."""

from __future__ import annotations

import json
import logging
import math
import os
import time
from typing import Any, Dict, List, Optional

from redis.exceptions import ResponseError


class OptimizedRedisCalculations:
    """
    Redis calculation helper that evaluates a Lua script via SCRIPT LOAD / EVALSHA.
    Falls back to the original Python implementations when Redis is unavailable.
    """

    def __init__(self, redis_client: Any):
        self.logger = logging.getLogger(__name__)
        self._script_sha: Optional[str] = None

        self.redis = self._resolve_client(redis_client)
        if not self.redis:
            self.logger.warning(
                "Redis client unavailable; falling back to Python calculations."
            )
            return

        # Lua script is in the same directory as this file (redis_files/)
        script_dir = os.path.dirname(os.path.abspath(__file__))
        self._script_path = os.path.normpath(os.path.join(script_dir, "redis_calculations.lua"))
        self._load_script()

    # ------------------------------------------------------------------ #
    # Internal helpers
    # ------------------------------------------------------------------ #

    def _resolve_client(self, redis_client: Any) -> Optional[Any]:
        """Resolve a redis-py client capable of script_load / evalsha."""
        if redis_client is None:
            return None

        # RobustRedisClient exposes .redis_client for raw access
        if hasattr(redis_client, "redis_client") and redis_client.redis_client:
            return redis_client.redis_client

        # Some helpers expose get_client(db)
        if hasattr(redis_client, "get_client"):
            try:
                client = redis_client.get_client(0)
                if client:
                    return client
            except Exception as exc:  # pragma: no cover - defensive
                self.logger.debug("Failed to resolve Redis client: %s", exc)

        # Lastly, accept a raw redis-py client (with script_load capability)
        if hasattr(redis_client, "script_load") and hasattr(redis_client, "evalsha"):
            return redis_client

        return None

    def _load_script(self) -> None:
        """Load the Lua script into Redis and cache its SHA."""
        try:
            with open(self._script_path, "r", encoding="utf-8") as lua_file:
                script = lua_file.read()
        except FileNotFoundError:
            self.logger.warning("Lua script not found at %s", self._script_path)
            self._script_sha = None
            return
        except OSError as exc:
            self.logger.warning("Unable to read Lua script: %s", exc)
            self._script_sha = None
            return

        try:
            self._script_sha = self.redis.script_load(script)
            self.logger.info("✅ Redis indicator script loaded")
        except Exception as exc:  # pragma: no cover - defensive
            self._script_sha = None
            self.logger.warning("Could not load Lua script, using Python fallbacks: %s", exc)

    def _encode_arg(self, value: Any) -> Any:
        if isinstance(value, (list, dict)):
            return json.dumps(value)
        if isinstance(value, bool):
            return "1" if value else "0"
        return value

    def _call_script(self, command: str, *args: Any) -> Any:
        if not self.redis or not self._script_sha:
            raise RuntimeError("Lua script not loaded")

        encoded_args = [command] + [self._encode_arg(arg) for arg in args]

        try:
            result = self.redis.evalsha(self._script_sha, 0, *encoded_args)
        except ResponseError as exc:
            if "NOSCRIPT" in str(exc).upper():
                self.logger.warning("Redis script missing; reloading.")
                self._load_script()
                if not self._script_sha:
                    raise
                result = self.redis.evalsha(self._script_sha, 0, *encoded_args)
            else:
                raise
        if isinstance(result, (bytes, bytearray)):
            return result.decode("utf-8")
        return result

    # ------------------------------------------------------------------ #
    # Public API (mirrors legacy RedisCalculations interface)
    # ------------------------------------------------------------------ #

    def calculate_atr_redis(
        self,
        high_data: List[float],
        low_data: List[float],
        close_data: List[float],
        period: int = 14,
    ) -> float:
        try:
            result = self._call_script("ATR", high_data, low_data, close_data, period)
            return float(result)
        except Exception as exc:
            self.logger.debug("Redis ATR fallback: %s", exc)
            return self._calculate_atr_python(high_data, low_data, close_data, period)

    def calculate_ema_redis(self, prices: List[float], period: int = 20) -> float:
        try:
            result = self._call_script("EMA", prices, period)
            return float(result)
        except Exception as exc:
            self.logger.debug("Redis EMA fallback: %s", exc)
            return self._calculate_ema_python(prices, period)

    def calculate_ema_all_redis(self, prices: List[float]) -> Dict[str, float]:
        """Calculate all EMA windows (5, 10, 20, 50, 100, 200) in one Redis call"""
        try:
            result = self._call_script("EMA_ALL", prices)
            if isinstance(result, str):
                return json.loads(result)
            return result
        except Exception as exc:
            self.logger.debug("Redis EMA_ALL fallback: %s", exc)
            return self._calculate_ema_all_python(prices)

    def calculate_rsi_redis(self, prices: List[float], period: int = 14) -> float:
        try:
            result = self._call_script("RSI", prices, period)
            return float(result)
        except Exception as exc:
            self.logger.debug("Redis RSI fallback: %s", exc)
            return self._calculate_rsi_python(prices, period)

    def calculate_vwap_redis(self, tick_data: List[Dict[str, Any]]) -> float:
        try:
            result = self._call_script("VWAP", tick_data)
            return float(result)
        except Exception as exc:
            self.logger.debug("Redis VWAP fallback: %s", exc)
            return self._calculate_vwap_python(tick_data)


    def calculate_indicators_batch(
        self, symbol: str, price_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        if not price_data:
            return {}
        try:
            response = self._call_script(
                "BATCH",
                price_data.get("prices", []),
                price_data.get("highs", []),
                price_data.get("lows", []),
                price_data.get("volumes", []),
            )
            if isinstance(response, str):
                return json.loads(response or "{}")
            if isinstance(response, dict):
                return response
        except Exception as exc:
            self.logger.debug("Redis batch calculation fallback: %s", exc)
        return self._python_indicators_fallback(price_data)

    # ------------------------------------------------------------------ #
    # Python fallbacks (existing implementations)
    # ------------------------------------------------------------------ #


    def _calculate_vwap_python(self, tick_data: List[Dict[str, Any]]) -> float:
        total_volume = 0.0
        weighted_sum = 0.0
        for tick in tick_data or []:
            # Try different volume fields in order of preference
            volume = (float(tick.get("zerodha_cumulative_volume", 0) or 0) or
                     float(tick.get("bucket_incremental_volume", 0) or 0) or
                     float(tick.get("bucket_cumulative_volume", 0) or 0) or 0)
            price = float(tick.get("last_price", 0) or 0)
            if volume > 0 and price > 0:
                total_volume += volume
                weighted_sum += volume * price
        return weighted_sum / total_volume if total_volume > 0 else 0.0

    def _calculate_atr_python(
        self,
        high_data: List[float],
        low_data: List[float],
        close_data: List[float],
        period: int,
    ) -> float:
        if (
            len(high_data) < period + 1
            or len(low_data) < period + 1
            or len(close_data) < period + 1
        ):
            return 0.0

        true_ranges = []
        for i in range(1, len(high_data)):
            high = float(high_data[i])
            low = float(low_data[i])
            prev_close = float(close_data[i - 1])
            tr1 = high - low
            tr2 = abs(high - prev_close)
            tr3 = abs(low - prev_close)
            true_ranges.append(max(tr1, tr2, tr3))

        if len(true_ranges) < period:
            return 0.0

        return sum(true_ranges[-period:]) / period

    def _calculate_rsi_python(self, prices: List[float], period: int) -> float:
        if len(prices) < period + 1:
            return 50.0

        gains = 0.0
        losses = 0.0
        for i in range(1, period + 1):
            change = float(prices[i]) - float(prices[i - 1])
            if change > 0:
                gains += change
            else:
                losses -= change

        avg_gain = gains / period
        avg_loss = losses / period

        for i in range(period + 1, len(prices)):
            change = float(prices[i]) - float(prices[i - 1])
            if change > 0:
                avg_gain = (avg_gain * (period - 1) + change) / period
                avg_loss = (avg_loss * (period - 1)) / period
            else:
                avg_gain = (avg_gain * (period - 1)) / period
                avg_loss = (avg_loss * (period - 1) - change) / period

        if avg_loss == 0:
            return 100.0

        rs = avg_gain / avg_loss
        return 100 - (100 / (1 + rs))

    def _calculate_ema_python(self, prices: List[float], period: int) -> float:
        if not prices:
            return 0.0
        if len(prices) < period:
            return float(prices[-1])

        alpha = 2.0 / (period + 1.0)
        ema = float(prices[0])
        for price in prices[1:]:
            ema = float(price) * alpha + ema * (1 - alpha)
        return ema

    def _calculate_ema_all_python(self, prices: List[float]) -> Dict[str, float]:
        """Calculate all EMA windows using Python fallback"""
        periods = [5, 10, 20, 50, 100, 200]
        result = {}
        for period in periods:
            result[f"ema_{period}"] = self._calculate_ema_python(prices, period)
        return result


    def _python_indicators_fallback(self, price_data: Dict[str, Any]) -> Dict[str, Any]:
        prices = price_data.get("prices", []) or []
        highs = price_data.get("highs", []) or []
        lows = price_data.get("lows", []) or []
        volumes = price_data.get("volumes", []) or []

        close_values = prices or highs or lows

        result = {
            "ema_5": self._calculate_ema_python(prices, 5) if prices else 0.0,
            "ema_10": self._calculate_ema_python(prices, 10) if prices else 0.0,
            "ema_20": self._calculate_ema_python(prices, 20) if prices else 0.0,
            "ema_50": self._calculate_ema_python(prices, 50) if prices else 0.0,
            "ema_100": self._calculate_ema_python(prices, 100) if prices else 0.0,
            "ema_200": self._calculate_ema_python(prices, 200) if prices else 0.0,
            "rsi": self._calculate_rsi_python(prices, 14) if prices else 50.0,
            "atr": self._calculate_atr_python(highs, lows, close_values, 14)
            if highs and lows and close_values
            else 0.0,
            "vwap": self._calculate_vwap_python(
                [{"price": p, "volume": v} for p, v in zip(prices, volumes)]
            )
            if prices and volumes
            else 0.0,
        }
        return result


RedisCalculations = OptimizedRedisCalculations

__all__ = ["OptimizedRedisCalculations", "RedisCalculations"]

File: redis_files/redis_calculations.lua
-- redis_calculations.lua
-- Consolidated trading indicator calculations for Redis 8.x using EVAL scripts.

local function abs(x)
    return x < 0 and -x or x
end

local function max3(a, b, c)
    local m = a
    if b > m then m = b end
    if c > m then m = c end
    return m
end

local function fast_atr(highs, lows, closes, period)
    if #highs < period + 1 then return 0 end
    local sum = 0
    for i = 2, math.min(#highs, period + 1) do
        local high = highs[i] or 0
        local low = lows[i] or 0
        local prev_close = closes[i - 1] or 0
        local tr1 = high - low
        local tr2 = abs(high - prev_close)
        local tr3 = abs(low - prev_close)
        sum = sum + max3(tr1, tr2, tr3)
    end
    return sum / period
end

local function fast_ema(prices, period)
    if #prices < period then return prices[#prices] or 0 end
    local alpha = 2.0 / (period + 1.0)
    local ema = prices[1] or 0
    for i = 2, #prices do
        ema = (prices[i] or 0) * alpha + ema * (1 - alpha)
    end
    return ema
end

local function fast_ema_all_windows(prices)
    local periods = {5, 10, 20, 50, 100, 200}
    local result = {}
    for _, period in ipairs(periods) do
        result["ema_" .. period] = fast_ema(prices, period)
    end
    return result
end

local function fast_rsi(prices, period)
    if #prices < period + 1 then return 50 end

    local avg_gain, avg_loss = 0, 0
    for i = 2, period + 1 do
        local change = (prices[i] or 0) - (prices[i - 1] or 0)
        if change > 0 then
            avg_gain = avg_gain + change
        else
            avg_loss = avg_loss - change
        end
    end
    avg_gain = avg_gain / period
    avg_loss = avg_loss / period

    for i = period + 2, #prices do
        local change = (prices[i] or 0) - (prices[i - 1] or 0)
        if change > 0 then
            avg_gain = (avg_gain * (period - 1) + change) / period
            avg_loss = (avg_loss * (period - 1)) / period
        else
            avg_gain = (avg_gain * (period - 1)) / period
            avg_loss = (avg_loss * (period - 1) - change) / period
        end
    end

    if avg_loss == 0 then return 100 end
    local rs = avg_gain / avg_loss
    return 100 - (100 / (1 + rs))
end

local function fast_vwap(tick_data)
    local total_volume, price_volume_sum = 0, 0
    for i = 1, #tick_data do
        local tick = tick_data[i] or {}
        -- Try different volume fields in order of preference
        local volume = (tonumber(tick.zerodha_cumulative_volume) or 
                       tonumber(tick.bucket_incremental_volume) or 
                       tonumber(tick.bucket_cumulative_volume) or 
                       tonumber(tick.volume) or 0)
        local price = tonumber(tick.last_price) or tonumber(tick.price) or 0
        if volume > 0 and price > 0 then
            total_volume = total_volume + volume
            price_volume_sum = price_volume_sum + (volume * price)
        end
    end
    if total_volume == 0 then return 0 end
    return price_volume_sum / total_volume
end

local function fast_indicator_batch(prices, highs, lows, volumes)
    local result = {
        ema_5 = fast_ema(prices, 5),
        ema_10 = fast_ema(prices, 10),
        ema_20 = fast_ema(prices, 20),
        ema_50 = fast_ema(prices, 50),
        ema_100 = fast_ema(prices, 100),
        ema_200 = fast_ema(prices, 200),
        rsi = fast_rsi(prices, 14),
        atr = fast_atr(highs, lows, prices, 14),
        vwap = 0.0
    }

    local total_volume, price_volume_sum = 0, 0
    local limit = math.min(#prices, #volumes)
    for i = 1, limit do
        local price = tonumber(prices[i]) or 0
        local volume = tonumber(volumes[i]) or 0
        if price > 0 and volume > 0 then
            total_volume = total_volume + volume
            price_volume_sum = price_volume_sum + price * volume
        end
    end
    result.vwap = total_volume > 0 and (price_volume_sum / total_volume) or 0
    return result
end

local command = ARGV[1] or ''

if command == 'ATR' then
    local highs = cjson.decode(ARGV[2] or '[]')
    local lows = cjson.decode(ARGV[3] or '[]')
    local closes = cjson.decode(ARGV[4] or '[]')
    local period = tonumber(ARGV[5]) or 14
    return tostring(fast_atr(highs, lows, closes, period))

elseif command == 'EMA' then
    local prices = cjson.decode(ARGV[2] or '[]')
    local period = tonumber(ARGV[3]) or 20
    return tostring(fast_ema(prices, period))

elseif command == 'EMA_ALL' then
    local prices = cjson.decode(ARGV[2] or '[]')
    return cjson.encode(fast_ema_all_windows(prices))

elseif command == 'RSI' then
    local prices = cjson.decode(ARGV[2] or '[]')
    local period = tonumber(ARGV[3]) or 14
    return tostring(fast_rsi(prices, period))

elseif command == 'VWAP' then
    local tick_data = cjson.decode(ARGV[2] or '[]')
    return tostring(fast_vwap(tick_data))


elseif command == 'ATR_BATCH' then
    local highs = cjson.decode(ARGV[2] or '[]')
    local lows = cjson.decode(ARGV[3] or '[]')
    local closes = cjson.decode(ARGV[4] or '[]')
    local period = tonumber(ARGV[5]) or 14
    return tostring(fast_atr(highs, lows, closes, period))

elseif command == 'BATCH' then
    local prices = cjson.decode(ARGV[2] or '[]')
    local highs = cjson.decode(ARGV[3] or '[]')
    local lows = cjson.decode(ARGV[4] or '[]')
    local volumes = cjson.decode(ARGV[5] or '[]')
    return cjson.encode(fast_indicator_batch(prices, highs, lows, volumes))

end

return '0'
